alexa/directive.js

```diff
@@ -105,42 +103,50 @@ class AlexaDirective extends AlexaResponse {
    *  based of interface-specific properties latest item state from OH
    *  and then return a formatted response to the Alexa request
    *
-   * @param {Object} parameters     Additional parameters [header, payload, response] (optional)
+   * @param {Object} parameters     Additional parameters [header, payload, properties, response] (optional)
    */
   getPropertiesResponseAndReturn(parameters = {}) {
     // Use the property map defined interface names if this.interface not defined (e.g. reportState)
     const interfaceNames = this.interface ? [this.interface] : Object.keys(this.propertyMap);
     // Get list of all unique item objects part of interfaces
     const interfaceItems = this.propertyMap.getItemsByInterfaces(interfaceNames);
-    const promises = [];
-
-    interfaceItems.forEach((item) => {
-      promises.push(this.getItemState(item).then((result) => {
+    // Define get item state promises array, excluding items with defined state already
+    const promises = interfaceItems.reduce((promises, item) => promises.concat(
+      typeof item.state !== 'undefined' ? [] : this.getItemState(item).then((result) => {
         // Update item information in propertyMap object for each item capabilities
         item.capabilities.forEach((capability) => {
           this.propertyMap[capability.interface][capability.property].item = result;
         });
         return result;
-      }));
-    });
+      })
+    ), []);
     Promise.all(promises).then((items) => {
       // Throw error if one of the state item is set to 'NULL'
       if (items.find(item => item.state === 'NULL')) {
```

Comment:
So i know this is not part of this commit, but this line just bit me.  I had several items in a group, one of them had a NULL value,  but it should not have mattered as the action i was taking (disarm) did not rely on it, but the response failed anyways.  This took me a while to figure out.  Is this line needed ?  I commented it out and it's working now, although i have not tried a command that depends the the item with a null state.

Reply:
I actually had similar issues during my testing and was considering changing the logic but the problem with NULL value is that they can be sent back in the properties response since this is considered as a string value, or could end up being normalized to undefined and then the properties check below would also throw an error.

I am open for suggestions on how to handle this case. The goal obviously is not to send invalid values back. One solution I was thinking was to remove the invalid values (NaN number type, NULL string type or undefined type) from the response and just send the ones that are valid.

Also, my guess why you were getting that error was probably because you have set some of alarm properties in the security panel group and these are getting pulled for the properties response. The logs should show the list of items with the ones that were set to NULL.

Reply:
>Also, my guess why you were getting that error was probably because you have set some of alarm properties 

Correct, I had NULL value for the state of a "SecurityPanelController.burglaryAlarm" which was causing it grief.  Since the is not used for a arm/disarm response it would be good to not cause it to fail.

I worry about users trying to debug this kind of stuff.  I was able to go through the logs and see debug info, the vast majority of our users will not be able to do this.  It would be nice if the binding was a forgiving as possible

Also I have not played around with how Alexa responds with the different types of error code  here https://developer.amazon.com/docs/device-apis/alexa-errorresponse.html#error-type-values  but it might be nice to make use of some of these and document them in the readme if Alexa can respond differently and we can map that to certain conditions.

Reply:
> Also I have not played around with how Alexa responds with the different types of error code here https://developer.amazon.com/docs/device-apis/alexa-errorresponse.html#error-type-values but it might be nice to make use of some of these and document them in the readme if Alexa can respond differently and we can map that to certain conditions.

So I just pushed a change to this PR to improve error handling. That should cover most of the issues you saw. I also included a [troubleshooting guide](https://github.com/openhab/openhab-alexa/blob/3de977a7b98d1d933643f5b1373c3481f5cd3d67/USAGE.md#troubleshooting).

Let me know what you think. Maybe we should put it higher in the page? Basically, the properties for undefined `NULL` or uninitialized `UNDEF` item states aren't included in the response but will change the health connectivity status to `UNREACHABLE`. Also, when restricting the list of properties to include in response (e.g. arm/disarm requests), the item state for the other properties part of the same interface will now not be pulled and therefore remained undefined. This is why there is a difference between undefined and null item states to identify the ones that we are expecting a valid state.

Based on my testing, this allows commands to be acknowledged as long as the relevant item state is valid while other item configured under the same interface may be invalid. However, it will generate an error when asking a device status for supported interfaces.

So using your test scenario as an example, Alexa will properly acknowledge when requesting to arm/disarm the device but it will return an error when requesting for the state as the alarm properties will be missing. Hopefully that makes sense.